### PR TITLE
Converted to gem package

### PR DESCRIPTION
**Pull request by andrew-pickin-epi**
* Makefile to control build[/test]/publish
* Added workflow to publish
* Moved version into package.json so more easily read by make
* Gemspec fixes

### CODE REVIEW

**Comment by andrew-pickin-epi** [comment]
There is scope here for placing the unit tests within the CICD workflow. 
I haven't explicitly done this, as I'm unsure of the consequences for development.

File: json_rails_logger.gemspec [hunk 1]
```diff
@@ -1,16 +1,25 @@
 # frozen_string_literal: true
 
-require './lib/json_rails_logger/version'
```
Comment (ijdickinson) [inline json_rails_logger.gemspec:3]: This is an unnecessary change. I don't want to introduce Npm's `package.json` for the sole purpose of carrying version information that we already have an established pattern for.

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,44 @@
+.PHONY: auth clean gem publish test
+
+NAME?=$(shell jq .name package.json | sed -e 's/"//g' | cut -f2 -d/)
+OWNER?=$(shell jq .name package.json | sed -e 's/"//g' | cut -f1 -d/)
+VERSION?=$(shell jq .version package.json | sed -e 's/"//g')
```
Comment (ijdickinson) [inline Makefile:5]: Replace with:

```
VERSION=$(shell ruby -e 'require "./lib/json_rails_logger/version" ; puts JsonRailsLogger::VERSION')
```

File: json_rails_logger.gemspec [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
   s.name        = 'json_rails_logger'
-  s.version     = JsonRailsLogger::VERSION
+  s.version     = pkg['version']
   s.date        = '2021-01-06'
   s.summary     = 'JSON Rails Logger'
   s.description = 'A custom rails logger that outputs JSON instead of raw text'
   s.authors     = ['Bogdan-Adrian Marc']
   s.email       = 'bogdan.marc@epimorphics.com'
-  s.files       = ['./lib/json_rails_logger.rb']
+  s.files       = [
```
Comment (ijdickinson) [inline json_rails_logger.gemspec:14]: A better solution is #27

File: lib/json_rails_logger.rb [hunk 4]
```diff
@@ -10,7 +10,6 @@
 require_relative 'json_rails_logger/json_formatter.rb'
 require_relative 'json_rails_logger/error.rb'
 require_relative 'json_rails_logger/logger.rb'
-require_relative 'json_rails_logger/version.rb'
```
Comment (ijdickinson) [inline lib/json_rails_logger.rb:13]: Please restore

File: package.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (ijdickinson) [inline package.json:1]: Please remove. There is no need to coerce a Ruby project to dress up as a JS project

**Review by ijdickinson — CHANGES_REQUESTED** [review]
OK with the makefile, not OK with the `package.json` suggestion. I've given a better solution.

**Comment by andrew-pickin-epi** [comment]
Will revert version.rb

**Comment by andrew-pickin-epi** [comment]
Do you object to me putting Name and Owner into version.rb as losing package.json doesn't provide a single value for name?

File: json_rails_logger.gemspec (hunk 3 above)
Comment (andrew-pickin-epi) [reply json_rails_logger.gemspec:13]: yes.  are you expecting #27 to get merged in here?

Comment (ijdickinson) [reply json_rails_logger.gemspec:13]: Yes, but it hasn't been approved yet

Comment (andrew-pickin-epi) [reply json_rails_logger.gemspec:13]: I'd rather not create a dependency on #27. Get this approved as it and #27 applied subsequently.

Comment (ijdickinson) [reply json_rails_logger.gemspec:13]: Has been merged now (the approval was in my email backlog)

File: lib/json_rails_logger/version.rb [hunk 6]
```diff
@@ -1,6 +1,8 @@
 # frozen_string_literal: true
 
 module JsonRailsLogger
+  NAME = "json_rails_logger"
```
Comment (ijdickinson) [inline lib/json_rails_logger/version.rb:4]: This duplicates information that is already in the Gemspec. You can read it like this:

```shell
ruby -e "puts Gem::Specification::load('json_rails_logger.gemspec').name"
```

File: lib/json_rails_logger/version.rb [hunk 7]
```diff
@@ -1,6 +1,8 @@
...
+  OWNER = "epimorphics"
```
Comment (ijdickinson) [inline lib/json_rails_logger/version.rb:5]: We should change the `authors` of the gemspec to be Epimorphics (.gemspec allows the singular `author` as well as `authors`), and then we can read the value from the gemspec in the same way as the `name`, above

**Review by ijdickinson — COMMENTED** [review]
There's no need to duplicate values in the `version.rb` that already are in the `.gemspec`. Although I do agree that in the Gemspec we should change the `authors: [bogdan]` to `author: "Epimorphics Ltd"

File: lib/json_rails_logger/version.rb (hunk 7 above)
Comment (andrew-pickin-epi) [reply lib/json_rails_logger/version.rb:5]: The case is significant. It's the location of the github registry url , it's not just boiler plate. So it needs to be `epimorphics` not "Epimorphics Ltd". Are you suggesting having both?
```
s.author = 'epimorphics'
s.authors = 'Epimorphics Ltd'
```

File: lib/json_rails_logger/version.rb (hunk 6 above)
Comment (andrew-pickin-epi) [reply lib/json_rails_logger/version.rb:4]: If you want, but you might as well hard code the name in the make file here. Since the name of the package is in the spec filename. So you no longer have a single source of truth for the package name and I was trying to avoid putting the package name in the makefile, as the inconsistent use of hyphen and underscore tripped me up at least once.

File: lib/json_rails_logger/version.rb (hunk 7 above)
Comment (ijdickinson) [reply lib/json_rails_logger/version.rb:5]: OK, so:

```
# gemspec
s.author = 'Epimorphics'

# Makefile
OWNER?=$(shell ruby -e "puts Gem::Specification::load('json_rails_logger.gemspec').author.downcase")
```

File: lib/json_rails_logger/version.rb (hunk 6 above)
Comment (ijdickinson) [reply lib/json_rails_logger/version.rb:4]: I do indeed want. We have a single source of truth in the gemspec file. Duplicating information that is already present in the gemspec in other files, like `package.json` or `version.rb`, dilutes that without solving a problem.

**Comment by andrew-pickin-epi** [comment]
Closing this as it has been accidentally been polluted by my forked repo.
I'll reopen another pr.
